### ISSUE TITLE
-DPERL_Y2KWARN doesn't do what it should (Not OK: perl v5.7.1    +DEVEL10104 on i586-linux 2.2.16-2

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3988]
Migrated from [rt.perl.org#6995](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=6995) (status was 'resolved')

Searchable as RT6995$

**Comment by p5pRT** [comment]
### From pnewton@gmx.de



\-DPERL\_Y2KWARN doesn't warn on Y2K problems\. However\, it warns
in some other situations instead \-\- this is most obvious when
installing modules\, where you'll see warnings on test 19 a lot
of times\.

Now I tried to build bleadperl with \-DPERL\_Y2KWARN\, and the 
warnings caused Perl not to pass its test suite\. Firstly\, it
failed the test for the Y2K warning :\)\, but a couple of other tests
also failed due to side effects from that\.

The warnings "Possibly Y2K bug&#8203;: about to append an integer to
'19' at blurfl line xyz\."\, when I investigated\, showed\, that line
xyz of blurfl was something like this&#8203;:

&nbsp;    if \($result\) \{
&nbsp;        print "ok $n\\n";

and occur when $n \(the test number\) is 19\. This was certainly the
case for the op/magic\.t and pragma/constant\.t failures in this case\.

Occasionally\, I also get "\.\.\.at \.\./lib/Symbol\.pm"\, perhaps while
generating the nineteenth anonymous symbol\.

Here's a short test case which will give a warning in bleadperl&#8203;:

$ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "ok $a\\n"'
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.

\(The first assignment to $b is to suppress "used only once"\.\)

Here's the Terse output of that&#8203;:

$ \./perl \-Ilib \-MO=Terse \-e '$a = 19; $b = ""; $b \.= "ok $a\\n"'
LISTOP \(0x81135e0\) leave \[1\]
&nbsp;    OP \(0x8113428\) enter
&nbsp;    COP \(0x8188948\) nextstate
&nbsp;    BINOP \(0x8188920\) sassign
&nbsp;        SVOP \(0x8188900\) const  IV \(0x81053c0\) 19
&nbsp;        UNOP \(0x81888e0\) null \[15\]
&nbsp;            SVOP \(0x8175ff0\) gvsv  GV \(0x8110a04\) \*a
&nbsp;    COP \(0x8112cd0\) nextstate
&nbsp;    BINOP \(0x8112ca8\) sassign
&nbsp;        SVOP \(0x8112c88\) const  PV \(0x8110a28\) ""
&nbsp;        UNOP \(0x8112c68\) null \[15\]
&nbsp;            SVOP \(0x8113608\) gvsv  GV \(0x8110998\) \*b
&nbsp;    COP \(0x810e3f0\) nextstate
&nbsp;    BINOP \(0x8113400\) concat \[4\]
&nbsp;        UNOP \(0x8112d30\) null \[15\]
&nbsp;            SVOP \(0x8112d10\) gvsv  GV \(0x8110998\) \*b
&nbsp;        UNOP \(0x81133b8\) null \[67\]
&nbsp;            OP \(0x81133e0\) null \[3\]
&nbsp;            BINOP \(0x8113500\) concat \[2\]
&nbsp;                BINOP \(0x81134d8\) concat \[1\]
&nbsp;                    SVOP \(0x8113448\) const  PV \(0x81109d4\) "ok "
&nbsp;                    UNOP \(0x8113498\) null \[15\]
&nbsp;                        SVOP \(0x8113478\) gvsv  GV \(0x8110a04\) \*a
&nbsp;                SVOP \(0x81134b8\) const  PV \(0x81109f8\) "\\n"
\-e syntax OK

Here's a patch that will catch at least one manifestation of the bug&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- t/pragma/warn/pp_hot.orig   Wed May  2 15:26:57 2001
+++ t/pragma/warn/pp_hot        Tue May 15 10:35:06 2001
@@ -225,6 +225,8 @@
 $x     = "19" . $yy . "\n";
 $x     = "319$yy\n";
 $x     = "319" . $yy . "\n";
+$yy = 19;
+$x     = "ok $yy\n";
 no warnings 'y2k';
 $x     = "19$yy\n";
 $x     = "19" . $yy . "\n";

```
</details>
End of patch.

This should produce no additional warnings if things work correctly\.

With this patch\, \`\./perl pragma/warnings\.t\` produces this&#8203;:

PROG&#8203;:
\# pp\_hot\.c \[pp\_concat\]
use warnings 'y2k';
use Config;
BEGIN \{
&nbsp;    unless \($Config\{ccflags\} =~ /Y2KWARN/\) \{
&nbsp;        print "SKIPPED\\n\# perl not built with \-DPERL\_Y2KWARN";
&nbsp;        exit 0;
&nbsp;    \}
\}
my $x;
my $yy = 78;
$x     = "19$yy\\n";
$x     = "19" \. $yy \. "\\n";
$x     = "319$yy\\n";
$x     = "319" \. $yy \. "\\n";
$yy = 19;
$x     = "ok $yy\\n";
no warnings 'y2k';
$x     = "19$yy\\n";
$x     = "19" \. $yy \. "\\n";
EXPECTED&#8203;:
Possible Y2K bug&#8203;: about to append an integer to '19' at \- line 12\.
Possible Y2K bug&#8203;: about to append an integer to '19' at \- line 13\.
GOT&#8203;:
Possible Y2K bug&#8203;: about to append an integer to '19' at \- line 17\.
not ok 319

"\- line 17" refers to the line with "ok $yy\\n"\.

A couple more data points&#8203;:

$ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "ok $a"'
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
$ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "ok $a\\n"'
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
$ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "$a\\n"'
$ \./perl \-w \-e '$a = 19; $b = ""; $b \.= " $a\\n"'
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
$ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "2 $a\\n"'
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
$ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "2$a\\n"'
$ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "19$a\\n"'
$ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "19 $a\\n"'
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.

So there has to be something before the interpolated variable for
the warning to fire\. But it apparently is silent if the "something"
is a digit \(even \-\- or\, perhaps\, especially \-\- if what's before
the interpolated variable is '19'\)\.

Here's the end of the output of \`\./perl harness\` in \./t \(without the patch\)&#8203;:

Failed Test       Status Wstat Total Fail  Failed  List of Failed
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
lib/open3\.t               29  7424    22    5  22\.73%  18\-22
op/magic\.t                 9  2304    38   20  52\.63%  19\-38
pragma/constant\.t                     82    1   1\.22%  36
pragma/warnings\.t                    428    1   0\.23%  319
9 tests and 87 subtests skipped\.
Failed 4/352 test scripts\, 98\.86% okay\. 27/23427 subtests failed\, 99\.88% okay\.


And here are the portions of the output that concern the failed
tests&#8203;:

lib/open3\.\.\.\.\.\.\.\.\.\.\.ok 17/22Possible Y2K bug&#8203;: about to append an integer to '19' at \.\./lib/Symbol\.pm line 86\, \<READ> line 5\.
lib/open3\.\.\.\.\.\.\.\.\.\.\.dubious
&nbsp;        Test returned status 29 \(wstat 7424\, 0x1d00\)
DIED\. FAILED tests 18\-22
&nbsp;        Failed 5/22 tests\, 77\.27% okay

op/magic\.\.\.\.\.\.\.\.\.\.\.\.ok 18/38Dying on warning&#8203;: Possible Y2K bug&#8203;: about to append an integer to '19' at op/magic\.t line 14\.
op/magic\.\.\.\.\.\.\.\.\.\.\.\.dubious
&nbsp;        Test returned status 9 \(wstat 2304\, 0x900\)
DIED\. FAILED tests 19\-38
&nbsp;        Failed 20/38 tests\, 47\.37% okay

pragma/constant\.\.\.\.\.FAILED test 36
&nbsp;        Failed 1/82 tests\, 98\.78% okay

pragma/warnings\.\.\.\.\.ok 318/428PROG&#8203;:
\# pp\_hot\.c \[pp\_concat\]
use warnings 'y2k';
use Config;
BEGIN \{
&nbsp;    unless \($Config\{ccflags\} =~ /Y2KWARN/\) \{
&nbsp;        print "SKIPPED\\n\# perl not built with \-DPERL\_Y2KWARN";
&nbsp;        exit 0;
&nbsp;    \}
\}
my $x;
my $yy = 78;
$x     = "19$yy\\n";
$x     = "19" \. $yy \. "\\n";
$x     = "319$yy\\n";
$x     = "319" \. $yy \. "\\n";
no warnings 'y2k';
$x     = "19$yy\\n";
$x     = "19" \. $yy \. "\\n";
EXPECTED&#8203;:
Possible Y2K bug&#8203;: about to append an integer to '19' at \- line 12\.
Possible Y2K bug&#8203;: about to append an integer to '19' at \- line 13\.
GOT&#8203;:

pragma/warnings\.\.\.\.\.FAILED test 319
&nbsp;        Failed 1/428 tests\, 99\.77% okay



And here's the output from the individual tests&#8203;:

$ \./perl lib/open3\.t
1\.\.22
ok 1
ok 2
ok 3
\[snip\]
ok 15
ok 16
ok 17
Possible Y2K bug&#8203;: about to append an integer to '19' at \.\./lib/Symbol\.pm line 86\, \<READ> line 5\.

$ \./perl op/magic\.t
1\.\.38
ok 1
ok 2
ok 3
\[snip\]
ok 16
ok 17
ok 18
Dying on warning&#8203;: Possible Y2K bug&#8203;: about to append an integer to '19' at op/magic\.t line 14\.

$ \./perl pragma/constant\.t
1\.\.82
ok 1
ok 2
ok 3
\[snip\]
ok 33
ok 34
ok 35
not ok 36
\# 36 : unexpected warning
\# 36 : Possible Y2K bug&#8203;: about to append an integer to '19' at pragma/constant\.t line 31\.
ok 37
ok 38
ok 39
\[snip\]
ok 80
ok 81
ok 82

$ \./perl pragma/warnings\.t
1\.\.428
\# From pragma/warn/1global
ok 1
ok 2
ok 3
\[snip\]
ok 303
ok 304
ok 305
\# From pragma/warn/pp\_hot
ok 306
ok 307
ok 308
ok 309
ok 310
ok 311
ok 312
ok 313
ok 314
ok 315
ok 316
ok 317
ok 318
PROG&#8203;:
\# pp\_hot\.c \[pp\_concat\]
use warnings 'y2k';
use Config;
BEGIN \{
&nbsp;    unless \($Config\{ccflags\} =~ /Y2KWARN/\) \{
&nbsp;        print "SKIPPED\\n\# perl not built with \-DPERL\_Y2KWARN";
&nbsp;        exit 0;
&nbsp;    \}
\}
my $x;
my $yy = 78;
$x     = "19$yy\\n";
$x     = "19" \. $yy \. "\\n";
$x     = "319$yy\\n";
$x     = "319" \. $yy \. "\\n";
no warnings 'y2k';
$x     = "19$yy\\n";
$x     = "19" \. $yy \. "\\n";
EXPECTED&#8203;:
Possible Y2K bug&#8203;: about to append an integer to '19' at \- line 12\.
Possible Y2K bug&#8203;: about to append an integer to '19' at \- line 13\.
GOT&#8203;:

not ok 319
ok 320
ok 321
\# From pragma/warn/pp\_sys
ok 322
ok 323
\[snip\]
ok 426
ok 427
ok 428

<details><summary>Perl Info</summary>

```

Flags:
    category=install
    severity=none

Site configuration information for perl v5.7.1:

Configured by pne at Tue May 15 09:12:31 BST 2001.

Summary of my perl5 (revision 5.0 version 7 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.2.16-22, archname=i586-linux
    uname='linux penderel 2.2.16-22 #1 tue aug 22 16:16:55 edt 2000 i586 unknown '
    config_args='-Dusedevel -Accflags=-DPERL_Y2KWARN -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-DPERL_Y2KWARN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-DPERL_Y2KWARN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.0)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL10104


@INC for perl v5.7.1:
    lib
    /usr/local/lib/perl5/5.7.1/i586-linux
    /usr/local/lib/perl5/5.7.1
    /usr/local/lib/perl5/site_perl/5.7.1/i586-linux
    /usr/local/lib/perl5/site_perl/5.7.1
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.7.1:
    HOME=/home/pne
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/pne/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/X11R6/bin
    PERL_BADLANG (unset)
    SHELL=/bin/sh

```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> A couple more data points&#8203;:
> 
> $ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "ok $a"'
> Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
> $ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "ok $a\\n"'
> Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
> $ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "$a\\n"'
> $ \./perl \-w \-e '$a = 19; $b = ""; $b \.= " $a\\n"'
> Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
> $ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "2 $a\\n"'
> Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
> $ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "2$a\\n"'
> $ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "19$a\\n"'
> $ \./perl \-w \-e '$a = 19; $b = ""; $b \.= "19 $a\\n"'
> Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.

And a couple more&#8203;:

$ \./perl \-w \-e '$a = 19; "ok " \. $a \. "\\n"'
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
$ \./perl \-w \-e '$a = 19; "ok " \. $a'
Useless use of concatenation \(\.\) or string in void context at \-e line 1\.
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
$ \./perl \-w \-e '$a = 19; "" \. $a'
Useless use of concatenation \(\.\) or string in void context at \-e line 1\.
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
$ \./perl \-w \-e '$a = 19; "1" \. $a'
Useless use of concatenation \(\.\) or string in void context at \-e line 1\.
$ \./perl \-w \-e '$a = 19; "$a"'
Useless use of string in void context at \-e line 1\.

So explicit contatenation with the empty string will trigger the 
warning\, but plain stringification won't\. And explicitly concatenating a 
number again results in the number being suppressed\.

Cheers\,
Philip

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

&nbsp;    if \(TARG == right\) \{
&nbsp;        if \(left == right\) \{
&nbsp;            /\*  $right = $right \. $right; \*/
&nbsp;            STRLEN rlen;
&nbsp;            char \*rpv = SvPV\(right\, rlen\);

&nbsp;            sv\_catpvn\(TARG\, rpv\, rlen\);
&nbsp;        \}
&nbsp;        else /\* $right = $left  \. $right; \*/
&nbsp;            sv\_catsv\(TARG\, rcopy\);
&nbsp;    \}
&nbsp;    else \{
&nbsp;        if \(\!SvOK\(TARG\)\) /\* Avoid warning when concatenating to undef\. \*/
&nbsp;            sv\_setpv\(TARG\, ""\);
&nbsp;        /\* $other = $left \. $right; \*/
&nbsp;        /\* $left  = $left \. $right; \*/
&nbsp;        sv\_catsv\(TARG\, right\);
&nbsp;    \}

\#if defined\(PERL\_Y2KWARN\)
&nbsp;    if \(\(SvIOK\(right\) || SvNOK\(right\)\) && ckWARN\(WARN\_Y2K\)\) \{
&nbsp;        STRLEN n;
&nbsp;        char \*s = SvPV\(TARG\,n\);
&nbsp;        if \(n >= 2 && s\[n\-2\] == '1' && s\[n\-1\] == '9'
&nbsp;            && \(n == 2 || \!isDIGIT\(s\[n\-3\]\)\)\)
&nbsp;        \{
&nbsp;            Perl\_warner\(aTHX\_ WARN\_Y2K\, "Possible Y2K bug&#8203;: %s"\,
&nbsp;                        "about to append an integer to '19'"\);
&nbsp;        \}
&nbsp;    \}
\#endif

If I understand this correctly\, perl has already done the
concat \(sv\_cat\(?&#8203;:sv|pvn\)\) before it gets around to checking
for the Y2K problem\. I presume that this sets TARG to the
concatenated result\.

Then the checking code checks TARG to see whether it ends
with "19"\, not preceded by a digit \-\- but this can only happen
if the right\-hand side was and the left\-hand side did not end
in a digit\! Hm\, or maybe if the right\-hand side is "9" and the
left\-hand side ends in /\\D1$/? Yep&#8203;:

$ \./perl \-w \-e '$a = 9; "ok 1$a"'
Useless use of string in void context at \-e line 1\.
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
$ \./perl \-w \-e '$a = 9; 1 \. $a'
Useless use of concatenation \(\.\) or string in void context at \-e line 1\.
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.

Here's another patch to pp\_hot\, to go instead of my previous one&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- t/pragma/warn/pp_hot.orig   Wed May  2 15:26:57 2001
+++ t/pragma/warn/pp_hot        Tue May 15 12:13:08 2001
@@ -225,6 +225,10 @@
 $x     = "19" . $yy . "\n";
 $x     = "319$yy\n";
 $x     = "319" . $yy . "\n";
+$yy = 19;
+$x     = "ok $yy\n";
+$yy = 9;
+$x     = 1 . $yy;
 no warnings 'y2k';
 $x     = "19$yy\n";
 $x     = "19" . $yy . "\n";

```
</details>
End of patch.

I'm working on a patch to pp\_hot\.c\.

More later\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

&nbsp;    if \(TARG == right\) \{
&nbsp;        if \(left == right\) \{
&nbsp;            /\*  $right = $right \. $right; \*/
&nbsp;            STRLEN rlen;
&nbsp;            char \*rpv = SvPV\(right\, rlen\);

&nbsp;            sv\_catpvn\(TARG\, rpv\, rlen\);
&nbsp;        \}
&nbsp;        else /\* $right = $left  \. $right; \*/
&nbsp;            sv\_catsv\(TARG\, rcopy\);
&nbsp;    \}
&nbsp;    else \{
&nbsp;        if \(\!SvOK\(TARG\)\) /\* Avoid warning when concatenating to undef\. \*/
&nbsp;            sv\_setpv\(TARG\, ""\);
&nbsp;        /\* $other = $left \. $right; \*/
&nbsp;        /\* $left  = $left \. $right; \*/
&nbsp;        sv\_catsv\(TARG\, right\);
&nbsp;    \}

\#if defined\(PERL\_Y2KWARN\)
&nbsp;    if \(\(SvIOK\(right\) || SvNOK\(right\)\) && ckWARN\(WARN\_Y2K\)\) \{
&nbsp;        STRLEN n;
&nbsp;        char \*s = SvPV\(TARG\,n\);
&nbsp;        if \(n >= 2 && s\[n\-2\] == '1' && s\[n\-1\] == '9'
&nbsp;            && \(n == 2 || \!isDIGIT\(s\[n\-3\]\)\)\)
&nbsp;        \{
&nbsp;            Perl\_warner\(aTHX\_ WARN\_Y2K\, "Possible Y2K bug&#8203;: %s"\,
&nbsp;                        "about to append an integer to '19'"\);
&nbsp;        \}
&nbsp;    \}
\#endif

If I understand this correctly\, perl has already done the
concat \(sv\_cat\(?&#8203;:sv|pvn\)\) before it gets around to checking
for the Y2K problem\. I presume that this sets TARG to the
concatenated result\.

Then the checking code checks TARG to see whether it ends
with "19"\, not preceded by a digit \-\- but this can only happen
if the right\-hand side was and the left\-hand side did not end
in a digit\! Hm\, or maybe if the right\-hand side is "9" and the
left\-hand side ends in /\\D1$/? Yep&#8203;:

$ \./perl \-w \-e '$a = 9; "ok 1$a"'
Useless use of string in void context at \-e line 1\.
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.
$ \./perl \-w \-e '$a = 9; 1 \. $a'
Useless use of concatenation \(\.\) or string in void context at \-e line 1\.
Possible Y2K bug&#8203;: about to append an integer to '19' at \-e line 1\.

Here's another patch to pp\_hot\, to go instead of my previous one&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- t/pragma/warn/pp_hot.orig   Wed May  2 15:26:57 2001
+++ t/pragma/warn/pp_hot        Tue May 15 12:13:08 2001
@@ -225,6 +225,10 @@
 $x     = "19" . $yy . "\n";
 $x     = "319$yy\n";
 $x     = "319" . $yy . "\n";
+$yy = 19;
+$x     = "ok $yy\n";
+$yy = 9;
+$x     = 1 . $yy;
 no warnings 'y2k';
 $x     = "19$yy\n";
 $x     = "19" . $yy . "\n";

```
</details>
End of patch.

I'm working on a patch to pp\_hot\.c\.

More later\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> I'm working on a patch to pp\_hot\.c\.

OK\, here we go\.

For your Changes file&#8203;: Fix PERL\_Y2KWARN test to check before rather than after appending

The patch&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- pp_hot.c.orig       Tue May  1 21:51:03 2001
+++ pp_hot.c    Tue May 15 13:30:28 2001
@@ -156,6 +156,19 @@
     if (TARG != left)
        sv_setsv(TARG, left);

+#if defined(PERL_Y2KWARN)
+    if ((SvIOK(right) || SvNOK(right)) && ckWARN(WARN_Y2K) && SvOK(TARG)) {
+       STRLEN n;
+       char *s = SvPV(TARG,n);
+       if (n >= 2 && s[n-2] == '1' && s[n-1] == '9'
+           && (n == 2 || !isDIGIT(s[n-3])))
+       {
+           Perl_warner(aTHX_ WARN_Y2K, "Possible Y2K bug: %s",
+                       "about to append an integer to '19'");
+       }
+    }
+#endif
+
     if (TARG == right) {
        if (left == right) {
            /*  $right = $right . $right; */
@@ -174,19 +187,6 @@
        /* $left  = $left . $right; */
        sv_catsv(TARG, right);
     }
-
-#if defined(PERL_Y2KWARN)
-    if ((SvIOK(right) || SvNOK(right)) && ckWARN(WARN_Y2K)) {
-       STRLEN n;
-       char *s = SvPV(TARG,n);
-       if (n >= 2 && s[n-2] == '1' && s[n-1] == '9'
-           && (n == 2 || !isDIGIT(s[n-3])))
-       {
-           Perl_warner(aTHX_ WARN_Y2K, "Possible Y2K bug: %s",
-                       "about to append an integer to '19'");
-       }
-    }
-#endif

     SETTARG;
     RETURN;

```
</details>
End of patch.

You'll notice that I basically moved the Y2K checking block further up\.
It's now after the place where C\<TARG> is set to C\<left> if it isn't the
same as C\<left> already\, but before the actual appending\.

After simply moving the block up\, I get a failure at op/assignwarn\.t test 8;
this tests that C\< \{ my $x; $x \.= 1; \} > does not produce a warning
\(i\.e\. appending stuff to undef\)\, but it did with my patch\. I presume that
came from the SvPV\(TARG\,n\) if TARG was undef\.

Then I noticed that there was a bit of code a few lines further down that
said&#8203;:

&nbsp;        if \(\!SvOK\(TARG\)\) /\* Avoid warning when concatenating to undef\. \*/
&nbsp;            sv\_setpv\(TARG\, ""\);

\, so I assume that something like this could be used as a workaround\.
But if TARG is undef\, then it doesn't end in "19"\, so I can skip the test\.
Therefore\, I added an "&& SvOK\(TARG\)" to the end of the condition\.
Please check whether this is correct here\.

With this patch in place\, my perl passed all tests \(including the two
extra "testlets" I put in pragma/warn/pp\_hot and for which I submitted
a patch earlier\)\. The days of warnings on test 19 will be over\.

\(I thought of submitting a further patch which checks for "20" at the
end\, but then discarded that\, since whoever outputs "20101" by
simply adding \(localtime\)\[5\]\, or "201" for \(\(localtime\)\[5\] % 100\)\,
deserves what they get\. If they think that's correct\, I'm not going to
put in a warning to save them\.\)

Enjoy\!

Cheers\,
Philip

**Comment by p5pRT** [comment]
### From @nwc10

As we're past Y2K001 now\, how useful is PERL\_Y2KWARN?
\[Although I realise it's still too easy to make Y2K bugs due to gmtime and
friends returning years from 1900\, instead of the more sensible 0 or 1970\]

Nicholas Clark

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> On Tue\, May 15\, 2001 at 04&#8203;:31&#8203;:05PM \+0200\, Philip Newton wrote&#8203;:
>
> > For your Changes file&#8203;: Fix PERL\_Y2KWARN test to check before rather than after appending
> 
> As we're past Y2K001 now\, how useful is PERL\_Y2KWARN?

Probably not a whole lot\. And I don't know how many people actually 
built their perl with \-DPERL\_Y2KWARN; it didn't seem widely 
publicised\. I can see it in perl56delta\.pod and Changes5\.6 \(which 
probably not a whole lot of people read\); there's an example of the \-
A switch to Configure in INSTALL that shows \-DPERL\_Y2KWARN being 
added to ccflags \(but without comment\, even though perl56delta 
points the reader to INSTALL for more info\); and finally\, there's a 
paragraph at the end of README\.Y2K\, which I suppose people 
interested in Y2K would read\.

Still\, if it's part of Perl\, and officially documented\, it should at least 
work :\)\. On the other hand\, I suppose one could argue that the 
whole thing could be removed now\, since \-\- as you say \-\- we're well 
past Y2K by now and people writing new code with "19" in it should 
be few\, and old code should have made it past code reviews\.

One of the main reasons I made the patch was because I was tired 
of warnings when running test scripts\.

> \[Although I realise it's still too easy to make Y2K bugs due to gmtime and
> friends returning years from 1900\, instead of the more sensible 0 or 1970\]

Yes\, ick\. Would have been a lot better\, probably\, if gmtime and 
friends had returned years from 0 right from the start\. The only 
advantage I can see for "year\-1900" is that it gives you a two\-digit 
year that you can \(be tempted to\) use straight away\, and it brought 
lots of problems when Y2K rolled around :\)\. With "year\-0"\, they would 
have had to use year%100 right from the start and wouldn't have 
had to change \(much; bad code would have resulted in this year 
being 201 instead of 19101\, presumably\)\. But it's a bit late to change 
that now\, I fear\.

\(And I read somewhere that POSIX defines the return value of time\(\) 
as the number of seconds since 1970\, as calculated from a day of 
86400 seconds exactly \[no leap seconds\] and leap years occurring 
every four years \-\- so in 2100 things are going to blow up \*again\*\, 
even \[or\, perhaps\, especially\] if you write standards\-conforming code\.\)

Cheers\,
Philip

**Comment by p5pRT** [comment]
### From @jhi

\(That the Y2K test had stopped working might have been my fault;
 I recently rewrote the append code from scratch\.\)

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> Thanks\, applied\.

Thanks\. Did you apply both the pp\_hot\.c patch and the two new test 
cases?

Cheers\,
Philip

**Comment by p5pRT** [comment]
### From @jhi

Rats\, missed the two new test cases\.

> Cheers\,
> Philip
> \-\- 
> Philip Newton \<pnewton@&#8203;gmx\.de>
> I appreciate copies of replies to my messages to Perl5 lists\.

**Comment by p5pRT** [comment]
### From @jhi

Now in\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> > Rats\, missed the two new test cases\.
> 
> Now in\.

Thank you\.

Cheers\,
Philip
\(Never hurts to have too many test cases\)
